presto-pinot-toolkit/src/main/java/com/facebook/presto/pinot/query/PinotFilterExpressionConverter.java

```diff
@@ -89,15 +100,50 @@ private PinotExpression handleLogicalBinary(
         }
         List<RowExpression> arguments = call.getArguments();
         if (arguments.size() == 2) {
-            return derived(format(
-                    "(%s %s %s)",
-                    arguments.get(0).accept(this, context).getDefinition(),
-                    operator,
-                    arguments.get(1).accept(this, context).getDefinition()));
+            String def0 = arguments.get(0).accept(this, context).getDefinition();
+            String def1 = arguments.get(1).accept(this, context).getDefinition();
+            def1 = handleTimeStampAndDateCast(arguments.get(0), def1);
+            def0 = handleTimeStampAndDateCast(arguments.get(1), def0);
+            return derived(format("(%s %s %s)", def0, operator, def1));
         }
         throw new PinotException(PINOT_UNSUPPORTED_EXPRESSION, Optional.empty(), format("Unknown logical binary: '%s'", call));
     }
 
+    private String handleTimeStampAndDateCast(RowExpression rowExpression, String definition)
+    {
+        if (!(rowExpression instanceof CallExpression)) {
+            return definition;
+        }
+        CallExpression callExpr = (CallExpression) rowExpression;
+        if (!standardFunctionResolution.isCastFunction(callExpr.getFunctionHandle())) {
+            return definition;
+        }
+        if (callExpr.getArguments().size() != 1) {
+            return definition;
+        }
+        Type inputType = callExpr.getArguments().get(0).getType();
+        Type expectedType = callExpr.getType();
+        if (inputType == DateType.DATE && (expectedType == TimestampType.TIMESTAMP || expectedType == TimestampWithTimeZoneType.TIMESTAMP_WITH_TIME_ZONE)) {
```

Comment:
Shouldn't we be returning output as a timestamp instead of number of days, if the expected type is a timestamp and input is a date ? We seem to be doing the opposite: returning the days when expected type is a timestamp.

Also what about timezone handling ?

Reply:
take example of `WHERE eventDate < current_time`.
`input type` is the `eventDate` field data type, which is `daysSinceEpoch`.
`expect type` is the right side literal type, which means right side is `millisSinceEpoch`. 
In order to make the pushdown work, we need to convert right side to `input type` which is date